### DISCUSSION TITLE
Subclassing MMM and overriding __init__ disrupts save/load functionality

### DISCUSSION THREAD

**Discussion by JnsLns** [pymc-labs/pymc-marketing#1219] [General]
Hi,

I am using a simple custom subclass of MMM, for which I have to execute just a bit of additional code in `__init__`, boiling down to this: 

```
class MMMSubclass(MMM):
    def __init__(self, *args, **kwargs):        
        super().__init__(*args, **kwargs)                
        print('Doing some additional stuff here...')
```

I noticed that this disupts the save/load functionality, because the `set_idata_attrs` method called at the end of `fit()`crashes with a value error: `ValueError: __init__ has parameters that are not in the attrs: {'args', 'kwargs'}. The save and load functionality will not work correctly.` This ultimately stems from comparing the signature of `__init__` to the expected set of attributes, which does not include `args` and `kwargs`.

The thing is, everything works up to the final portion of fit, expect for the idata attributes being set. Hence they cannot be reinstated at model load time later. 

Yet, I think all that stands in the way for this to work is the Value Error that prevents the final line of `ModelBuilder.set_idata_attrs` from being executed: `idata.attrs = attrs`. So one solution may be to make this Value Error a warning instead. 

Edit: Using version 0.10.0

**Comment by williambdean** [comment]
Hi @JnsLns 
What are you trying to do? Is this related to this issue: #1218 ?

**Comment by williambdean** [comment]
You can subclass and override specific methods if you would like. Including those dealing with save and load. You don't need to override the __init__ for a subclass. The way save and load works at the moment requires parameters or you'll run into that ValueError

**Comment by JnsLns** [comment]
Hi again @wd60622 !

Thanks for the hint, I am aware that I can override everything and thereby definitely can make it work somehow.  

My post was supposed to be less of a question rather than a suggestion. 

The point i wanted to make was simply that the existing save/load mechanic makes it impossible (and I feel needlessly so) to _conveniently_ override `__init__ ` for some minor modifications, since it prevents piping through `args` and `kwargs` as is common when subclassing. 

(Not directly related to https://github.com/pymc-labs/pymc-marketing/issues/1218)

**Reply by williambdean** [reply]
Yes, I understand that sentiment. However, on our end we are hedging from our save and load functionality from breaking. The user should be able to subclass MMM but is just required to be precise with the args and kwargs that are used.
